pub/lib/dependabot/pub/helpers.rb

```diff
@@ -0,0 +1,134 @@
+# frozen_string_literal: true
+
+require "json"
+require "open3"
+require "digest"
+
+require "dependabot/errors"
+require "dependabot/shared_helpers"
+require "dependabot/pub/requirement"
+
+module Dependabot
+  module Pub
+    module Helpers
+      private
+
+      def dependency_services_list
+        JSON.parse(run_dependency_services("list"))["dependencies"]
+      end
+
+      def dependency_services_report
+        sha256 = Digest::SHA256.new
+        dependency_files.each do |f|
+          sha256 << f.path + "\n" + f.content + "\n"
+        end
+        hash = sha256.hexdigest
+
+        cache_file = "/tmp/report-#{hash}-pid-#{Process.pid}.json"
+        return JSON.parse(File.read(cache_file)) if File.file?(cache_file)
+
+        report = JSON.parse(run_dependency_services("report"))["dependencies"]
+        File.write(cache_file, JSON.generate(report))
+        report
+      end
+
+      def dependency_services_apply(dependency_changes)
+        run_dependency_services("apply", stdin_data: dependencies_to_json(dependency_changes)) do
+          dependency_files.map do |f|
+            updated_file = f.dup
+            updated_file.content = File.read(f.name)
```

Comment:
Maybe I'm missing something obvious, but what's the reason for duplicating the file and reading the contents of the original file into it? Could we not return the actual file instead?

Reply:
Hmm, I think we are trying to create a Dependabot::DependencyFile that is exactly as the original just with updated content without knowing about the details of Dependabot::DependencyFile. Alternatively we should do something like:

```suggestion
            updated_file = DependencyFile.new(
              name: f.name,
              content: File.read(f.name),
              directory: f.directory,
              support_file: f.support_file?
            )
```

This is more fragile to changes in DependencyFile, but they will probably be rare...

Reply:
Ah gotcha, we've read the file contents from disk already into memory, but they have changed on disk and we need to pull in the updated contents 👍 this approach seems good, I think it's the most concise way to write it so let's keep as is!